test/apps/network/udp_err.c

```diff
@@ -154,6 +154,32 @@ FN_TEST(bind)
 }
 END_TEST()
 
+FN_TEST(bind_reuseaddr)
+{
+	sk_addr.sin_port = htons(8081);
+	struct sockaddr *psaddr = (struct sockaddr *)&sk_addr;
+	socklen_t addrlen = sizeof(sk_addr);
+
+	int enable = 1;
+	int socket1 = CHECK(socket(PF_INET, SOCK_DGRAM | SOCK_NONBLOCK, 0));
+	int socket2 = CHECK(socket(PF_INET, SOCK_DGRAM | SOCK_NONBLOCK, 0));
+
+	CHECK(bind(socket1, psaddr, addrlen));
+
+	TEST_ERRNO(bind(socket2, psaddr, addrlen), EADDRINUSE);
+
+	CHECK(setsockopt(socket1, SOL_SOCKET, SO_REUSEADDR, &enable,
+			 sizeof(enable)));
```

Comment:
This line doesn't look useful in Asterinas because
 - we never read `options.reuseaddr` again _after_ a successful `bind()` system call;
 - we never check if the previous socket was bound with `SO_REUSEADDR` (we only check the option value of the current binding socket).

But since you wrote this, I assume it has some effect in Linux? Perhaps add a FIXME to note this behavior difference?

Reply:
You were right, the test is insufficient. I have written more tests about `SO_REUSEADDR`. The test showed differently in Asterinas compared to Linux when sk1 disabled `SO_REUSEADDR` and sk2 enabled `SO_REUSEADDR`.

I have to find a way to combine the `SO_REUSEADDR` of both sk1 and sk2.

Reply:
> I have to find a way to combine the `SO_REUSEADDR` of both sk1 and sk2.

The `OptionSet` seems unreachable in `iface` in the network implementation. I doubt whether this can be quickly resolved, as it seems to require a network refactor. 

Perhaps adding a FIXME and output a warning message when binding with `can_reuse=true` as a temporary fix is acceptable? Since current applications follow the Linux manual by setting all sockets with `can_reuse=true` to use this function.

Reply:
Yes, I think we cannot trivially fix this behavioral difference.

Adding a FIXME is fine, and I'm not even sure it's worth a warning, since I just checked the [man pages](https://man7.org/linux/man-pages/man7/socket.7.html), which says in the "NOTES" section:
```
       Linux will allow port reuse only with the SO_REUSEADDR option
       when this option was set both in the previous program that
       performed a bind(2) to the port and in the program that wants to
       reuse the port.  This differs from some implementations (e.g.,
       FreeBSD) where only the later program needs to set the
       SO_REUSEADDR option.  Typically this difference is invisible,
       since, for example, a server program is designed to always set
       this option.
```

It's definitely a FIXME for Asterinas since we want our system calls to be ABI compatible with Linux, but the effect of this behavior seems to be rather limited according to the man pages.

Reply:
And I think in the current implementation, the warning message is even printed while we're creating backlog sockets for a listening socket (which always sets `can_reuse`). This can be noisy.